bin/common.c

```diff
@@ -75,3 +77,119 @@ int key_log_callback(void *file, struct s2n_connection *conn, uint8_t *logline,
 
     return fflush((FILE *)file);
 }
+
+/* An inverse map from an ascii value to a hexidecimal nibble value
+ * accounts for all possible char values, where 255 is invalid value */
+static const uint8_t hex_inverse[256] = {
```

Comment:
Do we really need a 256-byte array to convert to hex? Have you looked at the existing S2N hex code?

This would also be our ?third? hex implementation. Alternatives:
* You need this for setting the secret. Maybe customers will too. Maybe we need a PSK API that takes the secret as a hex string instead of as raw bytes.
* Maybe we should make one of our existing hex methods accessible here. Not is marked public in s2n.h, but is marked S2N_API (or probably for clarity, just explicitly marked with the correct visibility). I'm not sure how that fits into our visibility story.

Reply:
Yes, we currently have 3 implementation of hex but they all rely on internal s2n structs and are specific to a stuffer/blob use-case: 

- [s2n_key_log_hex_encode](https://github.com/aws/s2n-tls/blob/fe8df74ddafb712d3141056107db7228b19b6ef7/tls/s2n_key_log.h#L24)
- [s2n_stuffer_read_hex](https://github.com/aws/s2n-tls/blob/5c800d5369b84daa0140da8468af7cb422ba0f44/tests/testlib/s2n_testlib.h#L24)
- [s2n_hex_string_to_bytes](https://github.com/aws/s2n-tls/blob/fe8df74ddafb712d3141056107db7228b19b6ef7/utils/s2n_blob.h#L47)

We need to cleanup these hex functions and provide only one underlying solution for hex encoding and decoding.

For this use-case, we require a hex decoder that does not rely on internal structs and currently there isn't one. Adding Issue:https://github.com/aws/s2n-tls/issues/2759  to address this cleanup. This implementation is a copy of s2n_hex_string_to_bytes implementation which is the only CBMC tested one:  https://github.com/aws/s2n-tls/blob/fe8df74ddafb712d3141056107db7228b19b6ef7/utils/s2n_blob.c#L83-L135